src/ott/solvers/nn/icnn.py

```diff
@@ -182,15 +174,11 @@ def _compute_identity_map(input_dim: int) -> Tuple[jnp.ndarray, jnp.ndarray]:
 
   @nn.compact
   def __call__(self, x: jnp.ndarray) -> float:
-    for i in range(self.num_hidden + 2):
-      if i == 0:
-        z = self.w_xs[i](x)
-      # apply both transform on hidden state and x
-      # x is one step ahead as there is one more hidden layer for x
-      else:
-        z = jnp.add(self.w_zs[i - 1](z), self.w_xs[i](x))
-      if i != 0 or i != self.num_hidden + 1:
-        z = self.act_fn(z)
+    z = self.act_fn(self.w_xs[0](x))
+    for i in range(self.num_hidden):
+      z = jnp.add(self.w_zs[i](z), self.w_xs[i + 1](x))
+      z = self.act_fn(z)
+    z += self.pos_def_potential(x)
```

Comment:
Yes, this change significantly helps: reverting this and re-running the notebook will result in a significantly worse result. I'm not sure what the best intuition/reason is for this and it could just be related to the initialization, but I think giving this potential's scalar value to the convex path at the beginning rather than the input makes it hard for the convex path to learn the right transformation of the input since it's never able to directly access it.

It could work to concatenate the quadratic potential's value to `z` in L177 so the following convex layers would have access to both. It may be adding too many parameters but we could also parameterize whether all of these are added (1. if the quadratic potential is used at the beginning or end of the network, or both, and 2. if the convex path initially accesses `x`)

Reply:
> the convex path to learn the right transformation of the input since it's never able to directly access it.

not sure I follow what you mean by "never" here? Do you mean some kind of vanishing gradient problem that prevents the linear part in the gradient (hence quadratic in the potential) to be meaningfully updated? If that's the case I think the intuition kind of makes sense! it's no longer a "proper" ICNN as you had defined in your paper, but this works :)

Reply:
I wasn't thinking about the vanishing gradient w.r.t. the quadratic potential (although that makes sense too), but was referring to how the previous version with the quadratic potential at the beginning transforms the input into a single scalar-valued potential. It seems like it could create a bottleneck because the following layers in the input-convex path need to base the input-convex path on transformations of the initial potential's scalar value rather than of the vector-valued input.